ethereum/closest_block.go

```diff
@@ -0,0 +1,95 @@
+package ethereum
+
+import (
+	"context"
+	"github.com/forta-network/forta-core-go/domain"
+	"github.com/forta-network/forta-core-go/utils"
+	"math/big"
+	"time"
+)
+
+// GetClosestBlock Obtains the block closest to the time given
+func GetClosestBlock(ctx context.Context, eth Client, activeTime time.Time) (*domain.Block, error) {
+
+	minBlockNumber := big.NewInt(0)
+	maxBlockNumber, err := eth.BlockNumber(ctx)
+	if err != nil {
+		return nil, err
+	}
+	closestBlockNumber := new(big.Int).Add(maxBlockNumber, minBlockNumber)
+	closestBlockNumber = closestBlockNumber.Div(closestBlockNumber, big.NewInt(2))
+	closestBlock, err := eth.BlockByNumber(ctx, closestBlockNumber)
+	if err != nil {
+		return nil, err
+	}
+
+	for minBlockNumber.Cmp(maxBlockNumber) <= 0 {
+		closestBlockTime, err := closestBlock.GetTimestamp()
+		if err != nil {
+			return nil, err
+		}
+
+		if closestBlockTime.Equal(activeTime) {
+			break
+		} else if closestBlockTime.After(activeTime) {
+			maxBlockNumber = maxBlockNumber.Sub(closestBlockNumber, big.NewInt(1))
+		} else {
+			minBlockNumber = minBlockNumber.Add(closestBlockNumber, big.NewInt(1))
+		}
+
+		closestBlockNumber = new(big.Int).Add(maxBlockNumber, minBlockNumber)
+		closestBlockNumber = closestBlockNumber.Div(closestBlockNumber, big.NewInt(2))
+		closestBlock, err = eth.BlockByNumber(ctx, closestBlockNumber)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return closestBlock, nil
+}
+
+// GetClosestBlockBefore Obtains the closest block only before the current active time
+func GetClosestBlockBefore(ctx context.Context, eth Client, activeTime time.Time) (*domain.Block, error) {
```

Comment:
trying to understand usage of the before/after methods. where would we need this in the rewards distribution code?

Reply:
This is not necessarily for the rewards distribution, but in a situation where we need to get blocks in period where in the boundary multiple blocks can be generated say minute x and minute y then we would need to probably take the first and last blocks in those boundaries e.g. 

- start minute M1: blocks B1, B2, B3, B4
- end minute M2: blocks B10, B11, B12, B13
to cover the (M1, M2) range, i guess one needs (B1, B13)